chain/testing.go

```diff
@@ -0,0 +1,243 @@
+	ticket := make([]byte, binary.Size(f.seq))
+	binary.BigEndian.PutUint64(ticket, f.seq)
+	f.seq++
+
+	// Sum weight of parents' parent weight, plus one for each parent.
+	parentWeight := types.Uint64(0)
+	for i := 0; i < parent.Len(); i++ {
+		parentWeight += parent.At(i).ParentWeight + 1
+	}
+
+	height := types.Uint64(0)
+	if parent.Defined() {
+		height = parent.At(0).Height + 1
+	}
+
+	b := &types.Block{
+		Ticket:          ticket,
+		Miner:           f.minerAddress,
+		ParentWeight:    parentWeight,
+		Parents:         parent.Key(),
+		Height:          height,
+		Messages:        []*types.SignedMessage{},
+		MessageReceipts: []*types.MessageReceipt{},
+		//StateRoot:       stateRoot,
+		//Proof PoStProof `json:"proof"`
+		//Timestamp Uint64 `json:"timestamp"`
+	}
+	// Nonce intentionally omitted as it will go away.
+
+	if build != nil {
+		build(&BlockBuilder{b})
+	}
+
+	// Compute state root from block.
+	var err error
+	b.StateRoot, err = f.stateBuilder.ComputeStateRoot(b)
+	require.NoError(f.t, err)
+
+	f.blocks[b.Cid()] = b
+	return b
+
+}
+
+///// Block builder /////
+
+// BlockBuilder mutates blocks as they are generated.
+type BlockBuilder struct {
+	block *types.Block
+}
+
+// SetTicket sets the block's ticket.
+func (bb *BlockBuilder) SetTicket(ticket []byte) {
+	bb.block.Ticket = ticket
+}
+
+// SetTimestamp sets the block's timestamp.
+func (bb *BlockBuilder) SetTimestamp(timestamp types.Uint64) {
+	bb.block.Timestamp = timestamp
+}
+
+// IncHeight increments the block's height, implying a number of null blocks before this one
+// is mined.
+func (bb *BlockBuilder) IncHeight(nullBlocks types.Uint64) {
+	bb.block.Height += nullBlocks
+}
+
+// AddMessage adds a message & receipt to the block.
+func (bb *BlockBuilder) AddMessage(msg *types.SignedMessage, rcpt *types.MessageReceipt) {
+	bb.block.Messages = append(bb.block.Messages, msg)
+	bb.block.MessageReceipts = append(bb.block.MessageReceipts, rcpt)
+}
+
+///// State builder /////
+
+// StateBuilder abstracts the computation of state root CIDs from the chain builder.
+type StateBuilder interface {
+	ComputeStateRoot(block *types.Block) (cid.Cid, error)
+}
+
+// FakeStateBuilder computes a fake state CID by hashing the CIDs of a block's parents and messages.
+type FakeStateBuilder struct {
+}
+
+// ComputeStateRoot computes a fake state root for `Block`.
+func (FakeStateBuilder) ComputeStateRoot(block *types.Block) (cid.Cid, error) {
+	var fakeState []cid.Cid
+	for it := block.Parents.Iter(); !it.Complete(); it.Next() {
+		fakeState = append(fakeState, it.Value())
+	}
+	for i := 0; i < len(block.Messages); i++ {
+		mCid, err := block.Messages[i].Cid()
+		if err != nil {
+			return cid.Undef, err
+		}
+		fakeState = append(fakeState, mCid)
+	}
+
+	return makeCid(fakeState)
+}
+
+///// Interface implementations /////
+
+// GetBlock returns the block identified by `c`.
+func (f *Builder) GetBlock(ctx context.Context, c cid.Cid) (*types.Block, error) {
```

Comment:
This doesn't seem to be needed by syncFetcher or TipSetProvider.  Is there another interface this is trying to satisfy?

Reply:
`BlockProvider` in traversal. I added another static check for it.